producer/writer/shard_writer.go

```diff
@@ -0,0 +1,258 @@
+	}
+}
+
+func (w *sharedShardWriter) Write(d producer.RefCountedData) {
+	w.mw.Write(d)
+}
+
+// This is not thread safe, must be called in one thread.
+func (w *sharedShardWriter) UpdateInstances(
+	instances []placement.Instance,
+	cws map[string]consumerWriter,
+) {
+	var (
+		newInstancesMap = make(map[string]struct{}, len(instances))
+		toBeDeleted     = w.instances
+	)
+	for _, instance := range instances {
+		id := instance.Endpoint()
+		newInstancesMap[id] = struct{}{}
+		if _, ok := toBeDeleted[id]; ok {
+			// Existing instance.
+			delete(toBeDeleted, id)
+			continue
+		}
+		// Add the consumer writer to the message writer.
+		w.mw.AddConsumerWriter(id, cws[id])
+	}
+	for id := range toBeDeleted {
+		w.mw.RemoveConsumerWriter(id)
+	}
+	w.instances = newInstancesMap
+}
+
+func (w *sharedShardWriter) Close() {
+	if !w.isClosed.CAS(false, true) {
+		return
+	}
+	w.mw.Close()
+}
+
+type replicatedShardWriter struct {
+	sync.RWMutex
+
+	shard          uint32
+	numberOfShards uint32
+	mPool          messagePool
+	opts           Options
+	logger         log.Logger
+
+	ackRouter      ackRouter
+	replicaID      uint32
+	messageWriters map[string]messageWriter
+	isClosed       bool
+}
+
+func newReplicatedShardWriter(
+	shard, numberOfShards uint32,
+	router ackRouter,
+	mPool messagePool,
+	opts Options,
+) shardWriter {
+	return &replicatedShardWriter{
+		shard:          shard,
+		numberOfShards: numberOfShards,
+		mPool:          mPool,
+		opts:           opts,
+		logger:         opts.InstrumentOptions().Logger(),
+		ackRouter:      router,
+		replicaID:      0,
+		messageWriters: make(map[string]messageWriter),
+		isClosed:       false,
+	}
+}
+
+func (w *replicatedShardWriter) Write(d producer.RefCountedData) {
+	w.RLock()
+	mws := w.messageWriters
+	w.RUnlock()
+	for _, mw := range mws {
+		mw.Write(d)
+	}
+}
+
+// This is not thread safe, must be called in one thread.
+func (w *replicatedShardWriter) UpdateInstances(
+	instances []placement.Instance,
+	cws map[string]consumerWriter,
+) {
+	// TODO: Schedule time after shardcutoff to clean up message writers that
+	// are already cutoff. Otherwise it will wait until next placement change
+	// to clean up.
+	var (
+		newMessageWriters = make(map[string]messageWriter, len(instances))
+		toBeClosed        []messageWriter
+		toBeAdded         = make(map[placement.Instance]consumerWriter, len(instances))
+		oldMessageWriters = w.messageWriters
+	)
+
+	for _, instance := range instances {
+		id := instance.Endpoint()
+		if mw, ok := oldMessageWriters[id]; ok {
+			newMessageWriters[id] = mw
+			// Existing instance, try to update cutover cutoff times.
+			w.updateCutoverCutoffNanos(mw, instance)
+			continue
+		}
+		// This is a new instance.
+		toBeAdded[instance] = cws[id]
+	}
+	for id, mw := range oldMessageWriters {
+		if _, ok := newMessageWriters[id]; ok {
+			// Still in the new placement.
+			continue
+		}
+		if instance, cw, ok := anyKeyValueInMap(toBeAdded); ok {
```

Comment:
Would it be more readable to just create a new message writer compared to updating the existing one in place?

Reply:
It needs to reuse the existing message writer as it may still hold unacked message in its queue and does not have a functioning consumer to process/ack them. The logic here addresses that.

Reply:
Okay can you add some comments? It's not clear why this is necessary by just looking at the code.

Reply:
sure